ROLLO Website

gwenchee opened the pull request:
This PR contains all the files for the ROLLO website. View the built website at: https://gwenchee.github.io/rollo/. Once this PR is merged, I will add a GitHub action to host the website on arfc.github.io/rollo instead. This PR addresses #8

gwenchee committed:
moved stuff

gwenchee committed:
update Make

gwenchee committed:
gh

gwenchee committed:
revert

gwenchee committed:
yml?

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
make

gwenchee committed:
Merge branch 'docs' of github.com:gwenchee/rollo into docs

gwenchee committed:
no jekyll

gwenchee committed:
Update .travis.yml

gwenchee committed:
make

gwenchee committed:
t Merge branch 'docs' of github.com:gwenchee/rollo into docs

gwenchee committed:
Update conf.py

gwenchee committed:
Update conf.py

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
Update .travis.yml

gwenchee committed:
try

gwenchee committed:
try

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Create install.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Create index.rst

gwenchee committed:
Create basic.rst

gwenchee committed:
Create fhr-slab.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Create license.rst

gwenchee committed:
Create index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
Update index.rst

gwenchee committed:
good

gwenchee committed:
make

gwenchee committed:
nice

gwenchee committed:
edits

gwenchee committed:
more writing:

gwenchee committed:
more docs

gwenchee committed:
more docs

gwenchee committed:
edits

gwenchee committed:
edits

gwenchee committed:
new logo

gwenchee committed:
new logo

gwenchee committed:
updates

gwenchee committed:
updates

gwenchee committed:
updates

gwenchee committed:
progress

gwenchee committed:
more docs

gwenchee committed:
Merge remote-tracking branch 'upstream/main' into docs

gwenchee committed:
travis

gwenchee committed:
progress

gwenchee committed:
updates to docs

gwenchee committed:
as is

gwenchee committed:
example notebooks

gwenchee committed:
more docs

gwenchee committed:
edits

gwenchee committed:
rm examples

gwenchee committed:
ackley example

gwenchee committed:
ackley example

gwenchee committed:
ackley example

gwenchee committed:
ackley fin

gwenchee committed:
binh and korn

gwenchee committed:
critical sphere

gwenchee committed:
more docs

gwenchee committed:
rm rollo build

gwenchee committed:
rm rollo build

gwenchee committed:
sphinx_rtd_theme

gwenchee committed:
sphinx_book_theme

gwenchee committed:
test

gwenchee committed:
test this

gwenchee committed:
test this

gwenchee committed:
test this

gwenchee committed:
test this

gwenchee committed:
test this

gwenchee committed:
test this

gwenchee committed:
test this

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
bye travis

gwenchee committed:
Merge remote-tracking branch 'upstream/main' into docs

gwenchee committed:
forgot to commit

gwenchee committed:
update output terminal description

gwenchee committed:
more theory

gwenchee committed:
more theory

gwenchee committed:
descr out txt

gwenchee committed:
pics

gwenchee committed:
progress

gwenchee committed:
logo
make html

gwenchee committed:
progress

gwenchee committed:
done cp

gwenchee committed:
cp

gwenchee committed:
done parllel

gwenchee committed:
notebook

gwenchee committed:
license link

pep8speaks commented:
Hello @gwenchee! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:











There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2022-07-18 20:28:29 UTC

gwenchee committed:
pep8

nsryan2 commented:
I'm going through the text now, the setup looks good (love Sphinx!). Something I wanted to note off the bat is that there are some parts of the website that are still under development, and, before merging, it would be a good idea to make issues for the sections that still need content (like How to Build ROLLO Documentation). Also, love the logo you ended up going with!

nsryan2 commented:
This is part of my review, I just didn't know where exactly to put it:
Also, something to think about is that you've got information on installation and stuff in the top-level readme and you can import that directly into your docs with `.. mdinclude:: ../README.md` so you don't have to update multiple places. You can do the same thing with a contributing document so that it can exist in your repo separate from the docs and in your docs!

nsryan2 commented on docs/source/index.rst:
```diff
@@ -0,0 +1,38 @@
+It essentially couples the `DEAP (Distributed Evolutionary Algorithms in Python) 
+<https://deap.readthedocs.io/en/master/>`_ evolutionary algorithm driver with 
+nuclear software.
+**ROLLO** is nuclear-software agnostic, and designed to easily couple to any software. 
```
`nuclear software` in the previous sentence is ambiguous. Combine these two sentences for conciseness

gwenchee replied:
done

nsryan2 commented on docs/source/install.rst:
```diff
@@ -0,0 +1,55 @@
+Install Guide
+===================
+
+The install guide outlines the steps to install **ROLLO** on your computer. 
```
```suggestion
The guide outlines the steps to install **ROLLO** on your computer. 
```

samgdotson replied:
```suggestion
The installation guide outlines the steps to install **ROLLO** on your computer. 
```

@nsryan2 your suggestion is fine, though "the" guide suggests a singular manual -- this is just for installing.

nsryan2 commented on docs/source/install.rst:
```diff
@@ -0,0 +1,55 @@
+ROLLO Installation with PyPI
+----------------------------
+The Python Package Index (PyPI) is a repository of software for the Python 
+programming language. Pip installing ROLLO is the most straightforward way to install 
```
```suggestion
programming language. Pip is the most straightforward way to install ROLLO.
```

nsryan2 commented on docs/source/install.rst:
```diff
@@ -0,0 +1,55 @@
+-----------------------------------------------
+All ROLLO source code is hosted on `Github <https://github.com/arfc/rollo/>`_. 
+You must first install the dependencies below individually to successfully use 
+ROLLO that is built from source. 
```
Is it possible to do this with the setup.py script?

samgdotson replied:
@gwenchee I'm not sure if Mac OS X is different (I doubt it is), but if you add a `requirements.txt` file. Users can simply do `pip install -r requirements.txt`, which installs all of the packages automatically. Alternatively, you can write a command in the guide for them to copy and paste. E.g.
```bash
pip install deap jinja2 numpy jsonschema pytest
```
which would accomplish the same thing.

nsryan2 replied:
Can confirm that works on Mac!

gwenchee replied:
Done!

nsryan2 commented on docs/source/motivation.rst:
```diff
@@ -0,0 +1,32 @@
+===================
+Motivation
+===================
+Evolutionary algorithms have been applied to nuclear reactor design optimization. 
```
```suggestion
```

nsryan2 commented on docs/source/motivation.rst:
```diff
@@ -0,0 +1,32 @@
+Motivation
+===================
+Evolutionary algorithms have been applied to nuclear reactor design optimization. 
+Previously, reactor designers individually customized available evolutionary algorithm 
```
```suggestion
Previously, nuclear reactor designers individually customized available evolutionary algorithm 
```

nsryan2 commented on docs/source/motivation.rst:
```diff
@@ -0,0 +1,32 @@
+optimization with nuclear software.
+
+Therefore, the motivation for **ROLLO** is to limit these inconveniences and facilitate
+using evolutionary algorithms for reactor design optimization.
```
```suggestion
the use of evolutionary algorithms for reactor design optimization.
```

nsryan2 commented on docs/source/motivation.rst:
```diff
@@ -0,0 +1,32 @@
+
+- Effective: **ROLLO** is well documented, tested, and version-controlled on `Github <https://github.com/arfc/rollo/>`_. 
+- Flexible: **ROLLO** is nuclear software agnostic. Users can vary any imaginable parameter with any nuclear software. **ROLLO** uses a templating method to edit the input file of the coupled software.
+- Parallel: Users have the two options for running ROLLO in parallel.
```
```suggestion
- Parallel: Users have two options for running **ROLLO** in parallel.
```

nsryan2 commented on docs/source/motivation.rst:
```diff
@@ -0,0 +1,32 @@
+- Effective: **ROLLO** is well documented, tested, and version-controlled on `Github <https://github.com/arfc/rollo/>`_. 
+- Flexible: **ROLLO** is nuclear software agnostic. Users can vary any imaginable parameter with any nuclear software. **ROLLO** uses a templating method to edit the input file of the coupled software.
+- Parallel: Users have the two options for running ROLLO in parallel.
+- Reproducible: Data from every ROLLO run saves into a unique `pickled file <https://docs.python.org/3/library/pickle.html>`_.
```
```suggestion
- Reproducible: Data from every **ROLLO** run saves into a unique `pickled file <https://docs.python.org/3/library/pickle.html>`_.
```

nsryan2 commented:
The image [docs/source/pics/ind-demonstrate.png](https://github.com/arfc/rollo/pull/53/files#diff-813d407ef75af3c82c2aca24d7b4f3be8a094cedf55ec2e2c784a61ede445bef) should probably just be a block of text. I'm not quite sure where it is in the docs, otherwise, I would've commented it in-line.

[docs/source/pics/rollo-classes.png](https://github.com/arfc/rollo/pull/53/files#diff-7b166bec55cbe4dd3f2175ef4b5c055cbc07bc32f3dabda0ff5c27cfa0816a0a) should probably also be text rather than an image incase something needs to change.

nsryan2 commented on docs/source/theory.rst:
```diff
@@ -0,0 +1,60 @@
+
+
+When the user runs a **ROLLO** input file, the `Executor` class drives **ROLLO**'s 
+execution from beginning to end. The Executor calls `InputValidation`` to parse 
```
```suggestion
execution from beginning to end. The Executor calls `InputValidation` to parse 
```

nsryan2 commented on docs/source/theory.rst:
```diff
@@ -0,0 +1,60 @@
+execution from beginning to end. The Executor calls `InputValidation`` to parse 
+the input file to ensure that the user defined all mandatory parameters and used the 
+correct formatting. Next, it initializes an `Evaluation` object based on the 
+``evaluators`` specifications in the input file. It uses the `Evaluation`` object to 
```
```suggestion
``evaluators`` specifications in the input file. It uses the `Evaluation` object to 
```

nsryan2 commented on docs/source/theory.rst:
```diff
@@ -0,0 +1,60 @@
+parameters and return the output parameters calculated by the evaluator software. 
+Next, it uses the `ToolboxGenerator` to create an initialized DEAP toolbox object 
+based on the input file's ``algorithm`` specifications. The `ToolboxGenerator` 
+object accepts the `Evaluation` object and registers it as the toolbox's 'evaluate' 
```
```suggestion
object accepts the `Evaluation` object and registers it as the toolbox's `evaluate`
```

nsryan2 commented on docs/source/usersguide/checkpoint_rollo.rst:
```diff
@@ -0,0 +1,105 @@
+The checkpoint file updates after each generation runs successfully. 
+If the **ROLLO** simulation fails half way through a generation, the user can
+restart the simulation from the previous generation. 
+Therefore, if you're running a ROLLO simulation with 10 generations and it fails 
```
```suggestion
Therefore, if you're running a **ROLLO** simulation with 10 generations and it fails 
```

nsryan2 commented on docs/source/usersguide/checkpoint_rollo.rst:
```diff
@@ -0,0 +1,105 @@
+If the **ROLLO** simulation fails half way through a generation, the user can
+restart the simulation from the previous generation. 
+Therefore, if you're running a ROLLO simulation with 10 generations and it fails 
+during generation 3, you can restart the ROLLO simulation from generation 2. 
```
```suggestion
during generation 3, you can restart the **ROLLO** simulation from generation 2. 
```

nsryan2 commented on docs/source/usersguide/checkpoint_rollo.rst:
```diff
@@ -0,0 +1,105 @@
+   * - Attribute 
+     - Description
+   * - ``fitness``
+     - fitness tuple holds the objective values. The sign refers to whether the objective is maximize or minimize. 
```
```suggestion
     - fitness tuple holds the objective values. The sign refers to whether the objective is to maximize or minimize. 
```

nsryan2 commented on docs/source/usersguide/checkpoint_rollo.rst:
```diff
@@ -0,0 +1,105 @@
+   * - ``num``
+     - reactor model index in generation
+   * - ``output`` 
+     - tuple of reactor model individual's output parameters
```
```suggestion
     - tuple of reactor model individual output parameters
```

nsryan2 commented on docs/source/usersguide/parallel_rollo.rst:
```diff
@@ -0,0 +1,64 @@
+
+Serial (none)
+=============
+The serial run mode runs each reactor model in each generation serially. 
```
```suggestion
The serial run mode runs each reactor model in each generation in series. 
```

nsryan2 commented on docs/source/usersguide/parallel_rollo.rst:
```diff
@@ -0,0 +1,64 @@
+===============
+The multiprocessing mode replaces the default map function with the 
+`multiprocessing_on_dill <https://pypi.org/project/multiprocessing_on_dill/>`_ map.
+The multiprocessing_on_dill map chops the iterable into several chunks which it 
```
```suggestion
The `multiprocessing_on_dill` map chops the iterable into several chunks which it 
```

nsryan2 commented on docs/source/usersguide/parallel_rollo.rst:
```diff
@@ -0,0 +1,64 @@
+Serial (none)
+=============
+The serial run mode runs each reactor model in each generation serially. 
+The serial run mode utilizes a map() function to run the nuclear software for each 
```
```suggestion
The serial run mode utilizes a `map()` function to run the nuclear software for each 
```

nsryan2 commented on docs/source/usersguide/parallel_rollo.rst:
```diff
@@ -0,0 +1,64 @@
+
+Multiprocessing 
+===============
+The multiprocessing mode replaces the default map function with the 
```
```suggestion
The multiprocessing mode replaces the default `map()` function with the 
```

nsryan2 commented on docs/source/usersguide/parallel_rollo.rst:
```diff
@@ -0,0 +1,64 @@
+The job control mode utilizes the Unix system's job control features to give users 
+more flexibility with parallelization setup. This flexibility enables parallelization 
+across distributed memory systems such as clusters and supercomputers.
+This mode does not use the map() function. 
```
```suggestion
This mode does not use the `map()` function. 
```

nsryan2 commented on docs/source/usersguide/run_rollo.rst:
```diff
@@ -0,0 +1,106 @@
+**ROLLO** will then run the ``input_script``, ``execute`` files, and 
+``output_script`` file. 
+All their respective output files will be in the individual directory as well. 
+For each run, **ROLLO** will output a txt file with the terminal output from the run. 
```
```suggestion
For each run, **ROLLO** will output a .txt file with the terminal output from the run. 
```

nsryan2 commented on docs/source/usersguide/run_rollo.rst:
```diff
@@ -0,0 +1,106 @@
+All their respective output files will be in the individual directory as well. 
+For each run, **ROLLO** will output a txt file with the terminal output from the run. 
+The txt file is named `evaluator name` _ `file type` _ `out.txt`. Thus, for evaluator 
+named OpenMC and the input script, the txt file will be named 
```
```suggestion
named OpenMC and the input script, the .txt file will be named 
```

nsryan2 commented on docs/source/usersguide/run_rollo.rst:
```diff
@@ -0,0 +1,106 @@
+
+ROLLO Terminal Outputs 
+======================
+After each ROLLO generation runs, ROLLO will output the following table with details about 
```
```suggestion
After each **ROLLO** generation runs, **ROLLO** will output the following table with details about 
```

nsryan2 reviewed (approved):
A few comments on the text, otherwise it looks good!

samgdotson commented on docs/source/about_us.rst:
```diff
@@ -0,0 +1,26 @@
+----------------------------
+
+----------------------------------
+Recommended publication for citing
```
Gwen, if you have time I would recommend submitting a paper to [JOSS](https://joss.theoj.org/). They're pretty lightweight articles that just explain the need and the software -- which you have basically done with your prelim/thesis.

gwenchee replied:
Thanks for the suggestion, Sam! I plan to write a long paper describing ROLLO and the AHTR optimization I did for my dissertation. I'll discuss with Madicken if a JOSS paper makes sense as well.

samgdotson commented on docs/source/install.rst:
```diff
@@ -0,0 +1,55 @@
+  :alt: rollo-logo
+
+===================
+Install Guide
```
```suggestion
Installation Guide
```

samgdotson commented on docs/source/install.rst:
```diff
@@ -0,0 +1,55 @@
+----------------------------
+The Python Package Index (PyPI) is a repository of software for the Python 
+programming language. Pip installing ROLLO is the most straightforward way to install 
+it. 
```
```suggestion
```

Use this with @nsryan2's suggestion.

samgdotson reviewed (changes requested):
Nathan had some good comments. These are extremely minor changes. The website looks phenomenal -- you've done an excellent job and should be very proud, @gwenchee!

yardasol commented:
@gwenchee this website looks great! I'll do a proper review tomorrow morning hopefully.

gwenchee commented:
Wow! Thanks all @nsryan2 @samgdotson @yardasol  for such fast reviews! I'll address your comments shortly. Also, in this PR I included example notebooks which you can view here: https://github.com/arfc/rollo/wiki/Example-Jupyter-Notebooks/ , let me know if you have any comments about these as well.

gwenchee committed:
update to README

gwenchee committed:
update to README

gwenchee committed:
Update docs/source/install.rst

Co-authored-by: Sam Dotson <44342873+samgdotson@users.noreply.github.com>

gwenchee committed:
Update docs/source/install.rst

Co-authored-by: Nathan Ryan <nathansean0@gmail.com>

gwenchee committed:
Update docs/source/motivation.rst

Co-authored-by: Nathan Ryan <nathansean0@gmail.com>

gwenchee committed:
Update docs/source/usersguide/parallel_rollo.rst

Co-authored-by: Nathan Ryan <nathansean0@gmail.com>

gwenchee committed:
map

gwenchee committed:
Update docs/source/usersguide/parallel_rollo.rst

Co-authored-by: Nathan Ryan <nathansean0@gmail.com>

gwenchee committed:
Update docs/source/usersguide/run_rollo.rst

Co-authored-by: Nathan Ryan <nathansean0@gmail.com>

gwenchee committed:
Apply suggestions from code review

Co-authored-by: Nathan Ryan <nathansean0@gmail.com>

gwenchee committed:
Apply suggestions from code review

Co-authored-by: Sam Dotson <44342873+samgdotson@users.noreply.github.com>

gwenchee committed:
update wording

gwenchee committed:
Merge branch 'docs' of github.com:gwenchee/rollo into docs

gwenchee committed:
move fig to code snippet

gwenchee committed:
github actions update

gwenchee committed:
move fig to table

gwenchee committed:
requirements txt

gwenchee committed:
requirements

gwenchee committed:
example update

gwenchee commented:
Thanks, @nsryan2 and @samgdotson! I have addressed all your inline comments. As per your suggestions, I added a `requirements.txt` to ease the dependency install process and updated the docs to reflect it. 

@nsryan2 : 
>  it would be a good idea to make issues for the sections that still need content (like How to Build ROLLO Documentation).

I created #53 and #54 to address this. 

> Also, something to think about is that you've got information on installation and stuff in the top-level readme and you can import that directly into your docs with `.. mdinclude:: ../README.md` so you don't have to update multiple places. You can do the same thing with a contributing document so that it can exist in your repo separate from the docs and in your docs!

This is a great suggestion, but I decided not to do this since I don't want to make the README too bloated with information. Instead, I provide a brief description about installing and running ROLLO in the README with links to the docs. 

I also updated the figures you mentioned into text. I was being lazy, so I'm glad you pointed them out :)

samgdotson committed:
Adds @nsryan2 comments

Co-authored-by: Nathan Ryan <nathansean0@gmail.com>

samgdotson reviewed (approved):
Thanks for addressing our comments!

yardasol commented on .github/workflows/build-website.yml:
```diff
@@ -0,0 +1,50 @@
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+        branch: gh-pages
+        force: true
+        directory: ./docs/build/html
```
instead of force pushing (not a great practice), consider using [actions-gh-pages](https://github.com/peaceiris/actions-gh-pages) to deploy the site for you

gwenchee replied:
This is great, thanks @yardasol , I implemented it.

yardasol commented on docs/source/devguide/build_docs.rst:
```diff
@@ -0,0 +1,9 @@
+
+================================
+How to Build ROLLO Documentation
+================================
```
Just add the commands you use in your actions workflow file.

gwenchee replied:
Done

yardasol commented on docs/source/index.rst:
```diff
@@ -0,0 +1,38 @@
+   devguide/index
+   publications
+   about_us
+   License <https://github.com/arfc/rollo/blob/main/LICENSE/>
```
I'd strongly recommend adding API docs so users and developers don't have to dive into source files to explore the code.

yardasol commented on docs/source/motivation.rst:
```diff
@@ -0,0 +1,31 @@
+Motivation
+===================
+Previously, nuclear reactor designers individually customized available evolutionary algorithm 
+packages for their reactor design optimization problems.
```
Previously to what? This is a bit vague, consider adding some citations or extra descriptions of previous efforts.

gwenchee replied:
I removed it instead.

yardasol commented on docs/source/motivation.rst:
```diff
@@ -0,0 +1,31 @@
+Reactor designers have to set up parallelization to use the genetic algorithm 
+optimization with nuclear software.
+
+Therefore, the motivation for **ROLLO** is to limit these inconveniences and facilitate
+the use of evolutionary algorithms for reactor design optimization.
```
Move this to the top of the motivation section

yardasol replied:
```suggestion
The motivation for **ROLLO** is to provide a user-friendly tool that helps reactor designers use evolutionary algorithms for reactor design optimization.
```

yardasol commented on docs/source/theory.rst:
```diff
@@ -0,0 +1,96 @@
+======================
+Theory and Methodology
+======================
+**ROLLO** essentially couples an evolutionary algorithm driver with nuclear 
+software, such as neutron transport and thermal-hydraulics codes, to optimize 
+nuclear reactor design. 
```
```suggestion
**ROLLO** provides a framework to couple an evolutionary algorithm driver with nuclear 
software, such as neutron transport and thermal-hydraulics codes, to optimize 
nuclear reactor design. 
```

yardasol commented on docs/source/theory.rst:
```diff
@@ -0,0 +1,96 @@
+
+.. list-table::
+   :widths: 20 80
+   :header-rows: 1
+
+   * - Class
+     - Description
+   * - `InputValidation`
+     - The `InputValidation` class contains methods to read and validate the JSON 
+       **ROLLO** input file to ensure the user defined all key parameters. If they did 
+       not, **ROLLO** raises an exception to tell the user which parameters are missing.
+   * - `Evaluation`
+     - DEAP's fitness evaluator (as mentioned in Section 4.1.1) requires an evaluation 
+       function to evaluate each individual's fitness values. The `Evaluation` class 
+       contains a method that creates an evaluation function that runs the nuclear 
+       software and returns the required fitness values defined in the input file.
+   * - `ToolboxGenerator`
+     - The `ToolboxGenerator` class initializes DEAP's toolbox and creator modules 
+       with genetic algorithm hyperparameters defined in the input file.
+   * - `Constraints`
+     - The `Constraints` class contains methods to initialize constraints defined in 
+       the input file and applies the constraints by removing individuals that do not 
+       meet the constraint.
+   * - `BackEnd`
+     - The `BackEnd` class contains methods to save genetic algorithm population 
+       results into a pickled checkpoint file and to restart a partially completed 
+       genetic algorithm from the checkpoint file.
+   * - `Algorithm`
+     - The `Algorithm` class contains methods to initialize and execute the genetic 
+       algorithm. It executes a general genetic algorithm framework that uses the 
+       hyperparameters defined in the `ToolboxGenerator`, applies constraints defined 
+       in `Constraints`, evaluates fitness values using the evaluation function 
+       produced by `Evaluation`, and saves all the results with `BackEnd`.
+   * - `Executor`
+     - The Executor class drives the **ROLLO** code execution with the following steps:
+       1) User input file validation with InputValidation, 
+       2) Evaluation function generation with Evaluation, 
+       3) DEAP toolbox initialization with ToolboxGenerator,
+       4) Constraint initialization with Constraints, 
+       5) Genetic algorithm execution with Algorithm
```
This is fine, but again, I'd strongly recommend having API docs available on website before it gets moved to arfc.github.io

yardasol commented on docs/source/usersguide/checkpoint_rollo.rst:
```diff
@@ -0,0 +1,125 @@
+Restarting Simulation
+=====================
+The checkpoint file updates after each generation runs successfully. 
+If the **ROLLO** simulation fails half way through a generation, the user can
```
```suggestion
If the **ROLLO** simulation fails during a generation, the user can
```

yardasol commented on docs/source/usersguide/checkpoint_rollo.rst:
```diff
@@ -0,0 +1,125 @@
+
+The checkpoint file holds the **ROLLO** simulation results and acts as a restart 
+file. Thus, if a **ROLLO** simulation ends prematurely, users can use the checkpoint 
+file to restart the code from the most recent population and continue the simulation.
```
```suggestion
file.
```

You can delete this since it's repeated immediately after

yardasol commented on docs/source/usersguide/checkpoint_rollo.rst:
```diff
@@ -0,0 +1,125 @@
+   * - ``all`` [``outputs``]
+     - list of population outputs, each population list contains a list of reactor model individual's output parameters
+
+Each reactor model individual is an Ind class type with attributes and is a simple 
```
```suggestion
Each reactor model individual is an :class:`Ind` class type with attributes and is a simple 
```

yardasol commented on docs/source/usersguide/run_rollo.rst:
```diff
@@ -0,0 +1,106 @@
+ROLLO Execution
+===============
+Users define the **ROLLO** genetic algorithm's number of ``generations`` and 
+``population size`` (more description in :ref:`input file algorithm section 
```
```suggestion
``pop_size`` (more description in :ref:`input file algorithm section 
```

yardasol commented on docs/source/usersguide/setup_rollo.rst:
```diff
@@ -0,0 +1,379 @@
+Variable names must be strings. 
+The following table describes each variable's input parameter sub-requirements: 
+
+.. list-table::
+   :widths: 25 25 30 20
+   :header-rows: 1
+
+   * - Input Parameter
+     - Type
+     - Description
+     - Mandatory?
+   * - ``min``
+     - float
+     - minimum value
+     - yes
+   * - ``max``
+     - float
+     - max value
+     - yes
```
In the future, it may make sense to move your JSON schema to an external file so you can import it here into the docs. It will reduce the amount of things you need to change if your JSON schema changes. You can do this using the `sphinx-jsonschema` package.

yardasol reviewed (approved):
Hi @gwenchee. It's a lot of work getting a website set up so I just want to recognize that I think you did an absolutely fantastic job on this. I have a few comments I'd like you to consider. See them below.

Also, please add python API docs for ROLLO to the website before moving it to arfc.github.io

gwenchee committed:
Update docs/source/motivation.rst

Co-authored-by: Olek <45364492+yardasol@users.noreply.github.com>

gwenchee committed:
Update docs/source/theory.rst

Co-authored-by: Olek <45364492+yardasol@users.noreply.github.com>

gwenchee committed:
Apply suggestions from code review

Co-authored-by: Olek <45364492+yardasol@users.noreply.github.com>

gwenchee committed:
python api rollo

gwenchee committed:
test peaceiris girhub actions

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
test

gwenchee committed:
build docs

gwenchee committed:
rm vague statement

gwenchee commented:
@yardasol Thanks for the review! Based on your comments, I made the following updates to the PR. 

1) I have implemented your excellent suggestion to add Python API. 
2) I addressed #54 by adding the commands in the build-website actions workflow file.
> In the future, it may make sense to move your JSON schema to an external file so you can import it here into the docs. It will reduce the amount of things you need to change if your JSON schema changes. You can do this using the `sphinx-jsonschema` package.
3) I added #56 and #57 to address this in a future PR. 



Here's the built website link again for convenience: https://gwenchee.github.io/rollo/

yardasol commented on .github/workflows/build-website.yml:
```diff
@@ -0,0 +1,44 @@
+    - name: make the sphinx docs
+      run: |
+        make -C docs clean 
+        # sphinx-apidoc -f -o docs/source . -H Test -e -t docs/source/_templates
```
```suggestion
```
You shouldn't need to run `make clean` since this is the base repo (with no build files in it). Also don't need a commented out line.

yardasol commented on .github/workflows/build-website.yml:
```diff
@@ -0,0 +1,44 @@
+        python3 -m pip install sphinx-jinja2-compat==0.2.0b1
+        python3 -m pip install sphinx-autoapi
+
+    - name: make the sphinx docs
```
```suggestion
    - name: Make the sphinx docs
```

yardasol commented on .github/workflows/build-website.yml:
```diff
@@ -0,0 +1,44 @@
+        # sphinx-apidoc -f -o docs/source . -H Test -e -t docs/source/_templates
+        make -C docs html
+
+    - name: Deploy
```
```suggestion
    - name: Deploy docs to github pages
```

yardasol commented on .github/workflows/build-website.yml:
```diff
@@ -0,0 +1,44 @@
+      if: ${{ github.ref == 'refs/heads/docs' }}
+      with:
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+        publish_dir: ./docs/build/html
```
```suggestion
        publish_dir: ./docs/build/html
        commit_message: "update docs at commit: ${{ github.event.head_commit.message }}"
```
Might be a good idea to add a commit so you can track the progression of the `gh-pages` branch in the source tree.

yardasol commented on .github/workflows/build-website.yml:
```diff
@@ -0,0 +1,44 @@
+      # If you're changing the branch from main, 
+      # also change the `main` in `refs/heads/main` 
+      # below accordingly.
+      if: ${{ github.ref == 'refs/heads/docs' }}
```
```suggestion
```
I don't think you need this if-statement, but if you've tested it and the docs only build with it present, then leave it in the workflow file.

yardasol commented on .github/workflows/build-website.yml:
```diff
@@ -0,0 +1,44 @@
+      max-parallel: 1
+
+    steps:
+    - uses: actions/checkout@v2
```
```suggestion
    - uses: actions/checkout@v3
```
Use newest version

yardasol commented on .github/workflows/build-website.yml:
```diff
@@ -0,0 +1,44 @@
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python 3.8
+      uses: actions/setup-python@v2
```
```suggestion
      uses: actions/setup-python@v4
```

yardasol commented on docs/source/devguide/build_docs.rst:
```diff
@@ -0,0 +1,35 @@
+How to Build ROLLO Documentation
+================================
+
+In order to build the documentation in the ``docs`` directory, you will need to have 
+the Sphinx third-party Python package and several Sphinx extentions. 
+The easiest way to install Sphinx and its extensions is via pip:
+
+.. code-block:: sh
+
+    pip install sphinx
+    pip install nbsphinx
+    pip install sphinx_rtd_theme
+    pip install sphinx_book_theme
+    pip install sphinx-toolbox --user
+    pip install sphinx-jinja2-compat==0.2.0b1
+    pip install sphinx-autoapi
+
+To build the documentation as a webpage, go to ``docs`` and run: 
+
+.. code-block:: sh
+
+    make html
+
+To build the documentation as a PDF, go to ``docs`` and run: 
+
+.. code-block:: sh
+
+    make latexpdf
```
great work!

yardasol commented on docs/source/index.rst:
```diff
@@ -0,0 +1,39 @@
+   usersguide/index
+   Examples  <https://github.com/arfc/rollo/wiki/Example-Jupyter-Notebooks/>
+   devguide/index
+   autoapi/index
```
:)

yardasol commented on docs/source/motivation.rst:
```diff
@@ -0,0 +1,28 @@
+Reactor designers have to set up parallelization to use the genetic algorithm 
+optimization with nuclear software.
+
+The motivation for **ROLLO** is to provide a user-friendly tool that helps reactor designers use evolutionary algorithms for reactor design optimization.
```
Make the sentence on line 20 the first sentence in the Motivation section.

gwenchee replied:
Done

yardasol commented on docs/source/motivation.rst:
```diff
@@ -0,0 +1,28 @@
+===================
+Motivation
+===================
+The evolutionary algorithm setup is highly customizable with an assortment of 
```
```suggestion
Evolutionary algorithm setup is highly customizable with an assortment of 
```

yardasol commented on docs/source/index.rst:
```diff
@@ -0,0 +1,39 @@
+
+**ROLLO** is an open-source Python package that applies evolutionary algorithm 
+techniques to optimize nuclear reactor design.  
+It essentially couples the `DEAP (Distributed Evolutionary Algorithms in Python) 
```
```suggestion
It couples the `DEAP (Distributed Evolutionary Algorithms in Python) 
```

yardasol reviewed (approved):
Thanks for addressing my comments @gwenchee. I saw a few more things in the second pass, but mostly all the changes look good.

I'd recommend adding formatting to variable, function, and class names in the docstrings for your classes and functions so they render with styling and can be easily differentiated from normal text on the web-hosted API docs. This is a lot of work though so may be better suited for a new PR.

gwenchee committed:
Apply suggestions from code review

Co-authored-by: Olek <45364492+yardasol@users.noreply.github.com>

gwenchee committed:
move sentence

gwenchee commented:
Thanks @yardasol !

> I'd recommend adding formatting to variable, function, and class names in the docstrings for your classes and functions so they render with styling and can be easily differentiated from normal text on the web-hosted API docs. This is a lot of work though so may be better suited for a new PR.

I opened #58. 

The PR is ready to be merged @nsryan2 @samgdotson @yardasol
